File: projectNew/old_src/ActionStep.java
package com.f14.bg.listener;

import com.f14.F14bg.network.CmdFactory;
import com.f14.bg.GameMode;
import com.f14.bg.action.BgAction;
import com.f14.bg.action.BgResponse;
import com.f14.bg.exception.BoardGameException;
import com.f14.bg.player.Player;


public abstract class ActionStep<P extends Player, GM extends GameMode<P>> {
    protected boolean isOver;
    /**
     * 是否在结束时删除剩余的步骤
     */
    protected boolean clearOtherStep = false;
    /**
     * 步骤所属的监听器
     */
    protected ActionListener<P, GM> listener;

    /**
     * 步骤执行完成后是否自动结束
     *
     * @return
     */
    public boolean autoOver() {
        return true;
    }

    /**
     * 创建步骤结束时的消息
     *
     * @param gameMode
     * @param player
     * @return
     */

    protected BgResponse createStepOverResponse(GM gameMode, P player) {
        BgResponse res = CmdFactory.createGameResponse(this.getActionCode(), player.getPosition());
        res.setPublicParameter("stepCode", this.getStepCode());
        res.setPublicParameter("ending", true);
        return res;
    }

    /**
     * 创建步骤开始时的消息
     *
     * @param gameMode
     * @param player
     * @return
     */
    protected BgResponse createStepStartResponse(GM gameMode, P player) {
        BgResponse res = CmdFactory.createGameResponse(this.getActionCode(), player.getPosition());
        res.setPublicParameter("stepCode", this.getStepCode());
        return res;
    }

    /**
     * 步骤中的行动
     *
     * @param gameMode
     * @param action
     * @throws BoardGameException
     */
    protected abstract void doAction(GM gameMode, BgAction action) throws BoardGameException;

    /**
     * 执行步骤
     *
     * @param gameMode
     * @param action
     * @throws BoardGameException
     */
    public void execute(GM gameMode, BgAction action) throws BoardGameException {
        String stepCode = action.getAsString("stepCode");
        if (!this.getStepCode().equals(stepCode)) {
            throw new BoardGameException("你还不能执行这个行动!");
        }
        this.doAction(gameMode, action);
        if (this.autoOver()) {
            this.isOver = true;
        }
    }

    /**
     * 取得行动代码
     *
     * @return
     */
    public abstract int getActionCode();

    /**
     * 取得步骤所属的监听器
     *
     * @param <L>
     * @return
     */

    @SuppressWarnings("unchecked")
    public <L extends ActionListener<P, GM>> L getActionListener() {
        return (L) this.listener;
    }

    /**
     * 取得步骤代码
     *
     * @return
     */
    public abstract String getStepCode();

    /**
     * 判断步骤是否已经完成
     *
     * @return
     */
    public boolean isOver() {
        return this.isOver;
    }

    /**
     * 步骤结束时触发的事件
     *
     * @param gameMode
     * @param player
     * @throws BoardGameException
     */
    protected void onStepOver(GM gameMode, P player) throws BoardGameException {
        gameMode.getGame().sendResponse(player, this.createStepOverResponse(gameMode, player));
    }

    /**
     * 步骤开始时触发的事件
     *
     * @param gameMode
     * @param player
     * @throws BoardGameException
     */
    protected void onStepStart(GM gameMode, P player) throws BoardGameException {
        gameMode.getGame().sendResponse(player, this.createStepStartResponse(gameMode, player));
    }
}

File: projectNew/src/com/f14/bg/listener/ActionStep.kt
package com.f14.bg.listener

import com.f14.F14bg.network.CmdFactory
import com.f14.bg.BoardGameConfig
import com.f14.bg.action.BgAction.GameAction
import com.f14.bg.action.BgResponse
import com.f14.bg.exception.BoardGameException
import com.f14.bg.player.Player
import com.f14.bg.report.BgReport

/**
 * Created by 吹风奈奈 on 2017/7/19.
 */

abstract class ActionStep<P : Player, C : BoardGameConfig, R : BgReport<P>>(val listener: ActionListener<P, C, R>) {
    /**
     * 判断步骤是否已经完成
     */
    var isOver: Boolean = false
        protected set
    /**
     * 是否在结束时删除剩余的步骤
     */
    var clearOtherStep = false
        protected set

    /**
     * 步骤执行完成后是否自动结束
     */
    fun autoOver(): Boolean = true

    /**
     * 创建步骤结束时的消息
     * @param player
     * @return
     */
    protected fun createStepOverResponse(player: P): BgResponse {
        val res = CmdFactory.createGameResponse(this.actionCode, player.position)
        res.public("stepCode", this.stepCode)
        res.public("ending", true)
        return res
    }

    /**
     * 创建步骤开始时的消息
     * @param player
     * @return
     */
    protected open fun createStepStartResponse(player: P): BgResponse {
        val res = CmdFactory.createGameResponse(this.actionCode, player.position)
        res.public("stepCode", this.stepCode)
        return res
    }

    /**
     * 步骤中的行动
     * @param action
     * @throws BoardGameException
     */
    @Throws(BoardGameException::class)
    protected abstract fun doAction(action: GameAction)

    /**
     * 执行步骤
     * @param action
     * @throws BoardGameException
     */
    @Throws(BoardGameException::class)
    fun execute(action: GameAction) {
        val stepCode = action.getAsString("stepCode")
        if (this.stepCode != stepCode) {
            throw BoardGameException("你还不能执行这个行动!")
        }
        this.doAction(action)
        if (this.autoOver()) {
            this.isOver = true
        }
    }

    /**
     * 取得行动代码
     * @return
     */
    abstract val actionCode: Int

    /**
     * 取得步骤代码
     * @return
     */
    abstract val stepCode: String

    /**
     * 步骤结束时触发的事件
     * @param player
     * @throws BoardGameException
     */
    @Throws(BoardGameException::class)
    open fun onStepOver(player: P) {
        this.createStepOverResponse(player).send(listener.mode, player)
    }

    /**
     * 步骤开始时触发的事件
     * @param player
     * @throws BoardGameException
     */
    @Throws(BoardGameException::class)
    open fun onStepStart(player: P) {
        this.createStepStartResponse(player).send(listener.mode, player)
    }
}
